wcfsetup/install/files/ts/WoltLabSuite/Core/Controller/Popover.ts

```diff
@@ -0,0 +1,475 @@
+    if (data === undefined) {
+      throw new Error(`Unable to find element for object id '${objectId}' (identifier: '${identifier}').`);
+    }
+
+    let fragment = DomUtil.createFragmentFromHtml(content);
+    if (!fragment.childElementCount) {
+      fragment = DomUtil.createFragmentFromHtml("<p>" + content + "</p>");
+    }
+
+    data.content = fragment;
+    data.state = State.Ready;
+
+    if (this.activeId) {
+      const activeElement = this.elements.get(this.activeId)!.element;
+
+      if (activeElement.dataset.cacheId === cacheId) {
+        this.show();
+      }
+    }
+  }
+
+  /**
+   * Handles the mouse start hovering the popover-enabled element.
+   */
+  private mouseEnter(event: MouseEvent): void {
+    if (this.suspended) {
+      return;
+    }
+
+    if (this.timerEnter) {
+      window.clearTimeout(this.timerEnter);
+      this.timerEnter = undefined;
+    }
+
+    const id = DomUtil.identify(event.currentTarget as HTMLElement);
+    if (this.activeId === id && this.timerLeave) {
+      window.clearTimeout(this.timerLeave);
+      this.timerLeave = undefined;
+    }
+
+    this.hoverId = id;
+
+    this.timerEnter = window.setTimeout(() => {
+      this.timerEnter = undefined;
+
+      if (this.hoverId === id) {
+        this.show();
+      }
+    }, Delay.Show);
+  }
+
+  /**
+   * Handles the mouse leaving the popover-enabled element or the popover itself.
+   */
+  private mouseLeave(): void {
+    this.hoverId = "";
+
+    if (this.timerLeave) {
+      return;
+    }
+
+    this.timerLeave = window.setTimeout(() => this.hidePopover(), Delay.Hide);
+  }
+
+  /**
+   * Handles the mouse start hovering the popover element.
+   */
+  private popoverMouseEnter(): void {
+    if (this.timerLeave) {
+      window.clearTimeout(this.timerLeave);
+      this.timerLeave = undefined;
+    }
+  }
+
+  /**
+   * Shows the popover and loads content on-the-fly.
+   */
+  private show(): void {
+    if (this.timerLeave) {
+      window.clearTimeout(this.timerLeave);
+      this.timerLeave = undefined;
+    }
+
+    let forceHide = false;
+    if (this.popover.classList.contains("active")) {
+      if (this.activeId !== this.hoverId) {
+        this.hidePopover();
+
+        forceHide = true;
+      }
+    } else if (this.popoverContent.childElementCount) {
+      forceHide = true;
+    }
+
+    if (forceHide) {
+      this.popover.classList.add("forceHide");
+
+      // force layout
+      //noinspection BadExpressionStatementJS
+      this.popover.offsetTop;
+
+      this.clearContent();
+
+      this.popover.classList.remove("forceHide");
+    }
+
+    this.activeId = this.hoverId;
+
+    const elementData = this.elements.get(this.activeId);
+    // check if source element is already gone
+    if (elementData === undefined) {
+      return;
+    }
+
+    const cacheId = elementData.element.dataset.cacheId!;
+    const data = this.cache.get(cacheId)!;
+
+    if (data.state === State.Ready) {
```

Comment:
A `switch()` would be preferable for an `enum`. I believe the compiler can automatically detect if certain cases are not handled and those should get a comment why they do not need to be handled at that place.